File: src/main/java/org/frawa/elmtest/core/LabelUtils.java
package org.frawa.elmtest.core;

import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.io.FileUtil;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class LabelUtils {
    public static final String ELM_TEST_PROTOCOL = "elmTest";
    public static final String DESCRIBE_PROTOCOL = ELM_TEST_PROTOCOL + "Describe";
    private static final String TEST_PROTOCOL = ELM_TEST_PROTOCOL + "Test";
    public static final String ERROR_PROTOCOL = ELM_TEST_PROTOCOL + "Error";

    final static Path EMPTY_PATH = Paths.get("");

    private static String getModuleName(Path path) {
        return pathString(path.getName(0));
    }

    private static String encodeLabel(String label) {
        try {
            return URLEncoder.encode(label, "utf8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    static String decodeLabel(Path encoded) {
        try {
            return URLDecoder.decode(pathString(encoded), "utf8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    static Path toPath(List<String> labels) {
        List<String> encoded = labels.stream()
                .map(LabelUtils::encodeLabel)
                .collect(Collectors.toList());
        if (encoded.isEmpty()) {
            return EMPTY_PATH;
        }
        return Paths.get(
                encoded.get(0),
                encoded.subList(1, encoded.size()).toArray(new String[0])
        );
    }

    static String pathString(Path path) {
        return FileUtil.toSystemIndependentName(path.toString());
    }

    public static String getName(Path path) {
        return decodeLabel(path.getFileName());
    }

    static String toSuiteLocationUrl(Path path) {
        return toLocationUrl(DESCRIBE_PROTOCOL, path);
    }

    static String toTestLocationUrl(Path path) {
        return toLocationUrl(TEST_PROTOCOL, path);
    }

    private static String toLocationUrl(String protocol, Path path) {
        return String.format("%s://%s", protocol, pathString(path));
    }

    public static Pair<String, String> fromLocationUrlPath(String path) {
        Path path1 = Paths.get(path);
        String moduleName = getModuleName(path1);
        String moduleFile = String.format("tests/%s.elm", moduleName.replace(".", "/"));
        String label = path1.getNameCount() > 1 ? decodeLabel(path1.subpath(1, path1.getNameCount())) : "";
        return new Pair<>(moduleFile, label);
    }

    static Path commonParent(Path path1, Path path2) {
        if (path1 == null) {
            return EMPTY_PATH;
        }
        if (path1.getNameCount() > path2.getNameCount()) {
            return commonParent(path2, path1);
        }
        if (path2.startsWith(path1)) {
            return path1;
        } else {
            return commonParent(path1.getParent(), path2);
        }
    }

    static Stream<Path> subParents(Path path, Path excludeParent) {
        if (excludeParent == EMPTY_PATH) {
            // TODO remove duplication with below
            List<Path> result = new ArrayList<>();
            for (Path current = path.getParent(); current != null; current = current.getParent()) {
                result.add(current);
            }
            return result.stream();
        }

        if (!path.startsWith(excludeParent)) {
            throw new IllegalStateException("not parent");
        }

        if (path == EMPTY_PATH) {
            return Stream.empty();
        }

        List<Path> result = new ArrayList<>();
        for (Path current = path.getParent(); !current.equals(excludeParent); current = current.getParent()) {
            result.add(current);
        }
        return result.stream();

        // JSK 9
//        return Stream.iterate(path, current -> current != null ? current.getParent() : null)
//                .takeWile(current -> !current.equals(excludeParent));
    }

    static String toErrorLocationUrl(String path, int line, int column) {
        return String.format("%s://%s::%d::%d", ERROR_PROTOCOL, path, line, column);
    }

    public static Pair<String, Pair<Integer, Integer>> fromErrorLocationUrlPath(String spec) {
        String[] parts = spec.split("::");
        String file = parts[0];
        int line = parts.length > 1 ? Integer.parseInt(parts[1]) : 1;
        int column = parts.length > 2 ? Integer.parseInt(parts[2]) : 1;
        return new Pair<>(file, new Pair<>(line, column));
    }

}

File: src/main/kotlin/org/elm/ide/test/core/LabelUtils.kt
package org.elm.ide.test.core

import com.intellij.openapi.util.io.FileUtil
import org.elm.ide.test.core.LabelUtils.ERROR_PROTOCOL
import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.file.Path
import java.nio.file.Paths

object LabelUtils {

    val DESCRIBE_PROTOCOL = "elmTestDescribe"
    val TEST_PROTOCOL = "elmTestTest"
    val ERROR_PROTOCOL = "elmTestError"

    val EMPTY_PATH = Paths.get("")

    private fun getModuleName(path: Path): String {
        return pathString(path.getName(0))
    }

    private fun encodeLabel(label: String): String {
        return URLEncoder.encode(label, "utf8")
    }

    fun decodeLabel(encoded: Path): String {
        return decodeLabel(pathString(encoded))
    }

    fun decodeLabel(encoded: String): String {
        return URLDecoder.decode(encoded, "utf8")
    }

    fun toPath(vararg labels: String): Path {
        if (labels.isEmpty()) return EMPTY_PATH
        val encoded = labels.map { encodeLabel(it) }
        return Paths.get(encoded.first(), *encoded.drop(1).toList().toTypedArray())
    }

    fun pathString(path: Path): String {
        return FileUtil.toSystemIndependentName(path.toString())
    }

    fun getName(path: Path): String {
        return decodeLabel(path.fileName)
    }

    fun toLocationUrl(path: Path, isSuite: Boolean = false): String {
        val protocol = if (isSuite) DESCRIBE_PROTOCOL else TEST_PROTOCOL
        return String.format("%s://%s", protocol, pathString(path))
    }

    /**
     * @param testsRelativeDirPath The path to the directory containing the tests, relative to the project's root (i.e.
     * the folder containing `elm.json`). The supplied `path` is the location of the test file, relative to this directory.
     */
    fun fromLocationUrlPath(path: String, testsRelativeDirPath: String): Pair<String, String> {
        val path1 = Paths.get(path)
        val moduleName = getModuleName(path1)
        val moduleFile = "$testsRelativeDirPath/${moduleName.replace(".", "/")}.elm"
        val label = if (path1.nameCount > 1) decodeLabel(path1.subpath(1, path1.nameCount)) else ""
        return Pair(moduleFile, label)
    }

    fun commonParent(path1: Path?, path2: Path): Path {
        return when {
            path1 == null -> EMPTY_PATH
            path1.nameCount > path2.nameCount -> commonParent(path2, path1)
            path2.startsWith(path1) -> path1
            else -> commonParent(path1.parent, path2)
        }
    }

    fun subParents(path: Path, excludeParent: Path): Sequence<Path> {
        if (excludeParent === EMPTY_PATH) {
            var current: Path? = path
            return generateSequence {
                current = current?.parent
                current
            }
        }

        if (!path.startsWith(excludeParent)) {
            throw IllegalStateException("not parent")
        }

        if (path === EMPTY_PATH) {
            return sequenceOf()
        }

        var current: Path? = path
        return generateSequence {
            current = current?.parent
            if (current != excludeParent) {
                current
            } else {
                null
            }
        }
    }

}

data class ErrorLabelLocation(
        val file: String,
        val line: Int,
        val column: Int
) {
    fun toUrl() =
            String.format("%s://%s::%d::%d", ERROR_PROTOCOL, file, line, column)

    companion object {
        fun fromUrl(spec: String): ErrorLabelLocation {
            val parts = spec.split("::").dropLastWhile { it.isEmpty() }
            return ErrorLabelLocation(
                    file = parts[0],
                    line = if (parts.size > 1) Integer.parseInt(parts[1]) else 1,
                    column = if (parts.size > 2) Integer.parseInt(parts[2]) else 1
            )
        }
    }
}